src/OrchardCore.Modules/OrchardCore.Roles/Views/Workflows/Activities/GetUsersByRoleTask.cs

```diff
@@ -0,0 +1,79 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Localization;
+using OrchardCore.Users;
+using OrchardCore.Users.Models;
+using OrchardCore.Workflows.Abstractions.Models;
+using OrchardCore.Workflows.Activities;
+using OrchardCore.Workflows.Models;
+using OrchardCore.Workflows.Services;
+
+namespace OrchardCore.Roles.Workflows.Activities;
+
+public class GetUsersByRoleTask : TaskActivity
+{
+    private readonly UserManager<IUser> _userManager;
+    private readonly IWorkflowExpressionEvaluator _expressionEvaluator;
+    protected readonly IStringLocalizer S;
+
+    public GetUsersByRoleTask(UserManager<IUser> userManager, IWorkflowExpressionEvaluator expressionvaluator, IStringLocalizer<GetUsersByRoleTask> localizer)
+    {
+        _userManager = userManager;
+        _expressionEvaluator = expressionvaluator;
+        S = localizer;
+    }
+
+    public override string Name => nameof(GetUsersByRoleTask);
+
+    public override LocalizedString DisplayText => S["Get Users by Role Task"];
+
+    public override LocalizedString Category => S["User"];
+
+    public WorkflowExpression<string> OutputKeyName
+    {
+        get => GetProperty(() => new WorkflowExpression<string>());
+        set => SetProperty(value);
+    }
+
+    public IEnumerable<string> Roles
+    {
+        get => GetProperty(() => new List<string>());
+        set => SetProperty(value);
+    }
+
+    public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
+        => Outcomes(S["Done"], S["Failed"]);
+    
```

Comment:
Remove extra line here

Reply:
Yes. Not sure we should use Hashset. Maybe Dictionary<string, User> and user TryAdd

Storing the user is, would likely to force another query to the database. Also, if we store the user info on the state and someone restarts the workflow, they'll get the same info about the users as when the original workflow ran. But, let's see what Sebastien thinks. 

@sebastienros thoughts on storing the UserIds only in the workflow state vs storing the User object in the workflow state?

Reply:
@elaurentin after discussing with @sebastienros, we think there is a better way to store the Ids in the worflow state. Instead of storing the object (which is bad), or storing userIds (which has a consequence), store a dictionary of DocumentId and the userId. The reason document id is better is because YesSQL caches the User by document id "per request". This means is that the workflow will be slightly faster as it wont have to do another round trip to the database and fetch the users.

However, if you decide to store the document Id, you'll need to update the `UsersByIdFilter` to locate users by dictionary. For example change the `ProcessAsync` method in `UsersByIdFilter` class to start by something like this

```
if (input.Type == FluidValues.Dictionary)
{
    // Somehow convert the input to Dictionary<string, long>() Not sure the syntax without trying it.
    var items = input.Enumerate(ctx).SomeHowToDisctionary<long, string>();

    var cachedUsers = await _session.GetAsync<User>(items.Keys.ToArray());

    var cachedUserIds = cachedUsers.Select(x => x.UserId).ToHashSet();

    var missingUserIds = items.Values.Where(userId => !cachedUserIds.Contains(userId)).ToList();

    var missingUsers = await _session.Query<User, UserIndex>(x => x.UserId.IsIn(missingUserIds)).ListAsync();

    return FluidValue.Create(missingUsers.Concat(cachedUsers).ToList(), ctx.Options);
}

///.. Everything else will be kept the same.
```

Reply:
@elaurentin 
The problem with sending the UserIds is that you always forcing to query the database again. Once to generate the list and another to get the email in the same request which is not the best approach.

The idea of providing Dictionary<long, string> is that you can pass the document id as the key, and the user id as the value. Doing so will allow you to get the users by document Id "which will be returned from the cache instead of doing a second round trip to the database. 

This is way I suggested to add the new condition at the beginning of the filter implementation so that you can first attempt to get the users by the document ids.

Reply:
>I think we are talking about milliseconds

We are talking about possibly pulling a large list of record all over again not milliseconds. It could be milliseconds for your use case. But, in we need to implement something new in OC, we should try to keep it as clean as possible.

Anyway, I added he support to Dictionary as mentioned.